onnx/test/symbolic_shape_test.py

```diff
@@ -0,0 +1,135 @@
+from onnx.helper import make_model, make_node, make_tensor, make_tensor_value_info
+from typing import Sequence, Union, Text, Tuple, List, Any, Optional
+import onnx.shape_inference
+import unittest
+
+
+class TestSymbolicShape(unittest.TestCase):
+
+    def _assert_valueinfo_shape(self, onnx_model, value_infos):  # type: (ModelProto, List[ValueInfoProto]) -> None
+        """
+            Assert onnx_model.value_info should be the same as expected value_infos
+            Instead of exact symbol, use -1 to represent symbolic shape in expected value_infos
+        """
+        for expected_vi in value_infos:
+            shape = self._get_shape_from_name(onnx_model, expected_vi.name)
+            assert shape is not None, '{}'.format(onnx_model)
+            if expected_vi.type.HasField('tensor_type'):
+                expected_shape = expected_vi.type.tensor_type.shape
+            elif expected_vi.type.HasField('sparse_tensor_type'):
+                expected_shape = expected_vi.type.sparse_tensor_type.shape
+            assert len(shape.dim) == len(expected_shape.dim), '{}'.format(onnx_model)
+            for dim_i in range(len(shape.dim)):
+                dim = shape.dim[dim_i]
+                expected_dim = expected_shape.dim[dim_i]
+                # -1 means it's a symbolic shape
+                if expected_dim.dim_value == -1:
+                    # symbolic shape must exist
+                    assert dim.dim_param, '%s' % (onnx_model)
+                else:
+                    assert dim.dim_value == expected_dim.dim_value, '{}'.format(onnx_model)
+
+    def _count_unique_dim_param_number(self, onnx_model):  # type: (ModelProto) -> int
+        """
+           return the total number of unique symbolic shape
+        """
+        symbol_shape_set = set()
+        inputs = list(onnx_model.graph.input)
+        outputs = list(onnx_model.graph.output)
+        valueinfos = list(onnx_model.graph.value_info)
+        for v in inputs + outputs + valueinfos:
+            for dim in v.type.tensor_type.shape.dim:
+                if dim.dim_param:
+                    symbol_shape_set.add(dim.dim_param)
+        return len(symbol_shape_set)
+
+    def _get_shape_from_name(self, onnx_model, name):  # type: (ModelProto, Text) -> Optional[TensorShapeProto]
+        """
+            Get shape from tensor_type or sparse_tensor_type according to given name
+        """
+        inputs = list(onnx_model.graph.input)
+        outputs = list(onnx_model.graph.output)
+        valueinfos = list(onnx_model.graph.value_info)
+        for v in inputs + outputs + valueinfos:
+            if v.name == name:
+                if v.type.HasField('tensor_type'):
+                    return v.type.tensor_type.shape
+                elif v.type.HasField('sparse_tensor_type'):
+                    return v.type.sparse_tensor_type.shape
+        return None
+
+    def test_concat_enable_symbolic(self):  # type: () -> None
```

Comment:
can you add a test case for unknown shape... in all these tests there is a symbol present... can we add something where 

(!has_dim_value() && !has_dim_param()) == true